manager/orchestrator/task.go

```diff
@@ -67,7 +67,31 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
 		return false
 	}
 
-	return !reflect.DeepEqual(s.Spec.Task, t.Spec) ||
+	// Make shallow copy of the service for the comparison.
+	service := *s
+
+	// For non-failed tasks with a container spec runtime that have already
+	// pulled the required image (i.e., current state is between READY and
+	// RUNNING inclusively), ignore the value of the `PullOptions` field by
+	// setting the copied service to have the same PullOptions value as the
+	// task. A difference in only the `PullOptions` field should not cause
+	// a running (or ready to run) task to be considered 'dirty' when we
+	// handle updates.
+	// See https://github.com/docker/swarmkit/issues/971
+	currentState := t.Status.State
+	ignorePullOpts := api.TaskStateReady <= currentState && currentState <= api.TaskStateRunning
```

Comment:
I'd add `&& t.DesiredState <= api.TaskStateRunning`

This will make sure you're not ignoring pull options for a task that the orchestrator has decided shouldn't run anymore (and therefore, might need to be replaced if pull options are updated), for example if the node where it was running failed. `t.Status.State` is self-reported by the node, so when a node fails, this field can be stale for some time.

Generally the orchestrator only looks at `DesiredState`, except when detecting a task failure and setting `DesiredState` to `Shutdown`. This gives us a single source of truth and avoids making decisions based on out-of-date information. But I see what you're trying to do here.